src/AMAMMHOOK.sol

```diff
@@ -98,27 +173,53 @@ contract AMAMMHOOK is BaseHook {
         // if fee is on output, get the absolute output amount
         if (swapAmount < 0) swapAmount = -swapAmount;
 
-        IAmAmm.Bid memory _bid = _getLastManager(key.toId());
+        PoolId poolId = key.toId();
+        uint40 currentEpoch = _getEpoch(poolId, block.timestamp);
+        IAmAmm.Bid memory _bid = getLastManager(poolId, currentEpoch);
         uint24 fee = _getFee(_bid);
         address bidder = _bid.bidder;
         uint128 rent = _bid.rent;
-
         uint256 feeAmount = (uint128(swapAmount) * uint128(fee)) / TOTAL_BIPS;
         // manager takes fee
         console.log("feeAmount: ", feeAmount);
         poolManager.take(feeCurrency, bidder, feeAmount);
         // LP charge rent
         console.log("rent: ", rent);
-        // poolManager.take(bidToken, address(AMAMM), rent);
-
+        if(rent > 0) {
+            uint40 last = _lastChargedEpoch[poolId];
+            if(last == 0) {
+                last = _lastUpdatedEpoch[poolId] - 1; // -1 to charge from the last epoch
+            }
+            if (currentEpoch - last <= K(poolId)) {
+                rent = rent * uint128(currentEpoch - last);
+            } else {
+                rent = rent * uint128(K(poolId));
+            }
```

Comment:
> @leckylao have you considered the case where there is a gap in between epochs where there is no winner? In this logic it seems like it always recharges the bidder from getLastManager. I think there should be a case where nobody gets charged rent.
> 
> For exmaple: k=2 2-4 there is a bid. Current epoch is 4 and next bid is 6-8 ( lastupdatedepoch is 6-8). Between 4-6 we should not collect rent

That's the sandwich case, which we have put restriction on the future bid K range so that it would not happen.

Reply:
> Also what if there is only 1 bidder from 2-4 and no more bids after that. I think in this setup it will keep repeatedly charge the old bidder rent

That's why the condition is when rent > 0 from `getLastManager`, which in your case the getLastManager should return rent 0.

Reply:
ah okay makes sense.